storage/src/adb/current/ordered.rs

```diff
@@ -0,0 +1,1690 @@
+        let height = Self::grafting_height();
+        let grafted_mmr = GraftingStorage::<'_, H, _, _>::new(&self.status, mmr, height);
+        let mut proof = verification::range_proof(&grafted_mmr, start_loc..end_loc).await?;
+
+        // Collect the operations necessary to verify the proof.
+        let mut ops = Vec::with_capacity((*end_loc - *start_loc) as usize);
+        let futures = (*start_loc..*end_loc)
+            .map(|i| self.any.log.read(i))
+            .collect::<Vec<_>>();
+        try_join_all(futures)
+            .await?
+            .into_iter()
+            .for_each(|op| ops.push(op));
+
+        // Gather the chunks necessary to verify the proof.
+        let chunk_bits = BitMap::<H, N>::CHUNK_SIZE_BITS;
+        let start = *start_loc / chunk_bits; // chunk that contains the very first bit.
+        let end = (*end_loc - 1) / chunk_bits; // chunk that contains the very last bit.
+        let mut chunks = Vec::with_capacity((end - start + 1) as usize);
+        for i in start..=end {
+            let bit_offset = i * chunk_bits;
+            let chunk = *self.status.get_chunk_containing(bit_offset);
+            chunks.push(chunk);
+        }
+
+        let last_chunk = self.status.last_chunk();
+        if last_chunk.1 == 0 {
+            return Ok((proof, ops, chunks));
+        }
+
+        hasher.update(last_chunk.0);
+        proof.digests.push(hasher.finalize());
+
+        Ok((proof, ops, chunks))
+    }
+
+    /// Return true if the given sequence of `ops` were applied starting at location `start_loc` in
+    /// the log with the provided root.
+    pub fn verify_range_proof(
+        hasher: &mut Standard<H>,
+        proof: &Proof<H::Digest>,
+        start_loc: Location,
+        ops: &[Operation<K, V>],
+        chunks: &[[u8; N]],
+        root: &H::Digest,
+    ) -> bool {
+        super::verify_range_proof(
+            hasher,
+            Self::grafting_height(),
+            proof,
+            start_loc,
+            ops,
+            chunks,
+            root,
+        )
+    }
+
+    /// Generate and return a proof of the current value of `key`, along with the other
+    /// [KeyValueProofInfo] required to verify the proof. Returns KeyNotFound error if the key is
+    /// not currently assigned any value.
+    ///
+    /// # Errors
+    ///
+    /// Returns [Error::UncommittedOperations] if there are uncommitted operations.
+    /// Returns [Error::KeyNotFound] if the key is not currently assigned any value.
+    pub async fn key_value_proof(
+        &self,
+        hasher: &mut H,
+        key: K,
+    ) -> Result<(Proof<H::Digest>, KeyValueProofInfo<K, V, N>), Error> {
+        if self.status.is_dirty() {
+            return Err(Error::UncommittedOperations);
+        }
+        let op = self.any.get_key_loc(&key).await?;
+        let Some((value, next_key, loc)) = op else {
+            return Err(Error::KeyNotFound);
+        };
+        let height = Self::grafting_height();
```

Comment:
In `unordered.rs` we check for location validity at this point:

```rs
        if !loc.is_valid() {
            return Err(crate::mmr::Error::LocationOverflow(loc).into());
        }
```

I don't think we need this check -- this location is coming from the database, not an untrusted source. Should we remove this check from `unordered.rs` to keep the implementations closer?

Reply:
yes, removed.